[New Package] Node.js® v22: Initial v22.14.0

mreid-tt opened the pull request:
## Description

This PR includes the following:

1. Initial release of v22.14.0

Fixes #6460

## Checklist

- [x] Build rule `all-supported` completed successfully
- [x] New installation of package completed successfully
- [x] Package upgrade completed successfully (Manually install the package again)
- [x] Package [functionality was tested](https://github.com/SynoCommunity/spksrc/wiki/Package-Update-Policy#tests-checks)
- [x] Any needed [documentation](https://github.com/SynoCommunity/spksrc/wiki/Create-documentation) is updated/created


### Type of change

<!--Please use any relevant tags.-->
- [x] New Package

mreid-tt commented:
@hgy59, I'm not sure why the original request was closed, but I was able to build this for DSM 7.2. I successfully installed it on DSM 7.2 for x64, as shown in the log below. I also ran a basic functionality test, which I've included as well. Let me know your thoughts.

<details>
<summary>Install Log</summary>

```
2025/02/19 19:00:18	install nodejs_v22 22.14.0-1 Begin preinst
2025/02/19 19:00:18	Begin initialize_variables
2025/02/19 19:00:18	End initialize_variables
2025/02/19 19:00:19	===> Step preinst. STATUS=INSTALL USER=nodejs_v22 GROUP= SHARE_PATH=
2025/02/19 19:00:19	install nodejs_v22 22.14.0-1 End preinst ret=[0]
2025/02/19 19:00:23	install nodejs_v22 22.14.0-1 Begin /bin/mkdir -p /volume1/@appstore/nodejs_v22 -m 755
2025/02/19 19:00:23	install nodejs_v22 22.14.0-1 End /bin/mkdir -p /volume1/@appstore/nodejs_v22 -m 755 ret=[0]
2025/02/19 19:00:23	install nodejs_v22 22.14.0-1 Begin /bin/rm -rf /volume1/@appstore/nodejs_v22
2025/02/19 19:00:23	install nodejs_v22 22.14.0-1 End /bin/rm -rf /volume1/@appstore/nodejs_v22 ret=[0]
2025/02/19 19:00:23	install nodejs_v22 22.14.0-1 Begin /bin/mv -f /volume1/@tmp/synopkg/install.JuU9CF/package /volume1/@appstore/nodejs_v22
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 End /bin/mv -f /volume1/@tmp/synopkg/install.JuU9CF/package /volume1/@appstore/nodejs_v22 ret=[0]
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 Begin /bin/rm -rf /var/packages/nodejs_v22
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 End /bin/rm -rf /var/packages/nodejs_v22 ret=[0]
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 Begin /bin/mkdir -p /var/packages/nodejs_v22 -m 755
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 End /bin/mkdir -p /var/packages/nodejs_v22 -m 755 ret=[0]
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 Begin /bin/touch /var/packages/nodejs_v22/installing
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 End /bin/touch /var/packages/nodejs_v22/installing ret=[0]
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 Begin /bin/mv -f /volume1/@tmp/synopkg/install.JuU9CF/INFO /var/packages/nodejs_v22/INFO
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 End /bin/mv -f /volume1/@tmp/synopkg/install.JuU9CF/INFO /var/packages/nodejs_v22/INFO ret=[0]
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 Begin /bin/rm -rf /var/packages/nodejs_v22/scripts
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 End /bin/rm -rf /var/packages/nodejs_v22/scripts ret=[0]
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 Begin /bin/mv -f /volume1/@tmp/synopkg/install.JuU9CF/scripts /var/packages/nodejs_v22/scripts
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 End /bin/mv -f /volume1/@tmp/synopkg/install.JuU9CF/scripts /var/packages/nodejs_v22/scripts ret=[0]
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 Begin /bin/rm -rf /var/packages/nodejs_v22/WIZARD_UIFILES
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 End /bin/rm -rf /var/packages/nodejs_v22/WIZARD_UIFILES ret=[0]
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 Begin /bin/mv -f /volume1/@tmp/synopkg/install.JuU9CF/WIZARD_UIFILES /var/packages/nodejs_v22/WIZARD_UIFILES
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 End /bin/mv -f /volume1/@tmp/synopkg/install.JuU9CF/WIZARD_UIFILES /var/packages/nodejs_v22/WIZARD_UIFILES ret=[0]
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 Begin /bin/rm -rf /var/packages/nodejs_v22/conf
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 End /bin/rm -rf /var/packages/nodejs_v22/conf ret=[0]
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 Begin /bin/mv -f /volume1/@tmp/synopkg/install.JuU9CF/conf /var/packages/nodejs_v22/conf
2025/02/19 19:00:26	install nodejs_v22 22.14.0-1 End /bin/mv -f /volume1/@tmp/synopkg/install.JuU9CF/conf /var/packages/nodejs_v22/conf ret=[0]
2025/02/19 19:01:09	install nodejs_v22 22.14.0-1 Begin postinst
2025/02/19 19:01:09	Begin initialize_variables
2025/02/19 19:01:09	End initialize_variables
2025/02/19 19:01:09	===> Step postinst. STATUS=INSTALL USER=nodejs_v22 GROUP= SHARE_PATH=
2025/02/19 19:01:09	Begin save_wizard_variables
2025/02/19 19:01:09	End save_wizard_variables
2025/02/19 19:01:09	Begin syno_sync_var_folder
2025/02/19 19:01:10	End syno_sync_var_folder
2025/02/19 19:01:10	install nodejs_v22 22.14.0-1 End postinst ret=[0]
2025/02/19 19:01:13	install nodejs_v22 22.14.0-1 Begin start-stop-status start
2025/02/19 19:01:13	install nodejs_v22 22.14.0-1 End start-stop-status start ret=[0]
```
</details>

<details>
<summary>Test Log</summary>

```
mreid@DSM72:~$ node -v
v22.14.0
mreid@DSM72:~$ cat test_app.js 
var msg = 'Hello World';
console.log(msg);

mreid@DSM72:~$ node test_app.js 
Hello World
mreid@DSM72:~$ 
```
</details>

hgy59 commented:
> I'm not sure why the original request was closed,

I closed it, because it was addressed to synology.com.
Our latest nodejs package was v16. Then we dropped the creation of nodejs packages, because synology has (more or less) uptodate nodejs support.
I guess synology will add nodejs 22 sooner or later...

mitchoklemen commented:
Great! Can I test it?

where can I get v22 node.spk?

mreid-tt commented:
> I closed it, because it was addressed to synology.com.

Good catch — I overlooked this in my checks. I've reopened the issue and updated the request details to better align with the user's needs.

> Our latest nodejs package was v16. Then we dropped the creation of nodejs packages, because synology has (more or less) uptodate nodejs support. I guess synology will add nodejs 22 sooner or later...

It makes sense for us not to have a parallel deployment if Synology is handling this. However, similar to their PHP support, they don’t seem to include versions that are still in active development. If you're okay with it, I'd like to proceed with adding this.

I based my work on the previous v16 package, removed the older code, and it appears to compile correctly. However, I have a couple of uncertainties:  
1. The remaining `CONFIGURE_ARGS` in `cross/nodejs_v22/Makefile`  
2. Which binaries should be included in `cross/nodejs_v22/PLIST`  

If you could review these and provide some guidance, I'd really appreciate it.

mreid-tt commented:
> Great! Can I test it? where can I get v22 node.spk?

Yes, you can test it. First, determine your NAS's CPU architecture (refer to the Wiki). Then, go to the **Checks → Build** tabs above to find the relevant artifacts. Download the matching version, unzip it, and manually install it in DSM.

hgy59 commented:
> > Great! Can I test it? where can I get v22 node.spk?
> 
> Yes, you can test it. First, determine your NAS's CPU architecture (refer to the Wiki). Then, go to the **Checks → Build** tabs above to find the relevant artifacts. Download the matching version, unzip it, and manually install it in DSM.

I have recently added instructions on the FAQ page in the wiki on how to install packages from github build artifacts

mitchoklemen commented:
> > > Great! Can I test it? where can I get v22 node.spk?
> > 
> > 
> > Yes, you can test it. First, determine your NAS's CPU architecture (refer to the Wiki). Then, go to the **Checks → Build** tabs above to find the relevant artifacts. Download the matching version, unzip it, and manually install it in DSM.
> 
> I have recently added instructions on the FAQ page in the wiki on how to install packages from github build artifacts

I will try, but never done it before.
Thank you for your effort.

hgy59 commented:
> > I have recently added instructions on the FAQ page in the wiki on how to install packages from github build artifacts
> 
> I will try, but never done it before. Thank you for your effort.

this is the related section: https://github.com/SynoCommunity/spksrc/wiki/Frequently-Asked-Questions#download-development-and-pre-release-packages-created-by-github-build-action

mreid-tt commented:
@hgy59, what are your thoughts on this package? Should we proceed with publishing? If so, I'd appreciate your feedback on my queries above.

hgy59 commented:
> @hgy59, what are your thoughts on this package? Should we proceed with publishing? If so, I'd appreciate your feedback on my queries above.

- I would like to have feedback whether homebridge works with this (requested in https://github.com/SynoCommunity/spksrc/issues/6460)

- I expect a comment on the need to defined `CC_host` and `CXX_host` (it should not differ for x64 and other arch, since the host is always x64)

- the comment for DSM 7.2 requirement could be:
   ```
   # https://github.com/nodejs/node/blob/main/BUILDING.md
   # Required: kernel >= 4.181, glibc >= 2.28
   # C++ compiler: g++ >= 10.1.0 or clang++ >= 8.0.0
   # Host: GCC >= 12.2
   REQUIRED_MIN_DSM = 7.2
   ```

- @mitchoklemen confirmed it works on x64, but we should validate this package on armv7 and aarch64 systems too (a minimal test of the target arch of all binaries in the package can be done with the `file` command).

- If homebridge works, we should add a PR upstream to enhance the `install.sh` script for this new package (after we published it).

hgy59 commented:
@mreid-tt  `node` is the only binary file in the package and the target archs seem ok:
```
nodejs_v22_x64-7.2_22.14.0-1/package/bin/node: ELF 64-bit LSB pie executable, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, stripped
nodejs_v22_aarch64-7.2_22.14.0-1/package/bin/node: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, stripped
nodejs_v22_armv7-7.2_22.14.0-1/package/bin/node: ELF 32-bit LSB pie executable, ARM, EABI5 version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux-armhf.so.3, for GNU/Linux 3.2.0, stripped
```

mreid-tt commented:
@hgy59, since Synology has released a version, should we cancel this PR?

mreid-tt commented:
> @mreid-tt `node` is the only binary file in the package and the target archs seem ok

Does that mean the `PLIST` should only contain:
```
bin:bin/node
```

hgy59 commented:
> Does that mean the `PLIST` should only contain:

no, that is not the case
there are some lnk and a lot of js files.
we only could omit the "*.ps1" files (i.e. powershell scripts)

hgy59 commented:
> @hgy59, since Synology has released a version, should we cancel this PR?

Yes synology added node js 22 packages, so this PR is obsolete.
https://archive.synology.com/download/Package/Node.js_v22/22.13.1-1004

It is not (yet) release 22.14, but it is not worth the effort to publish competing nodejs 22 packages.
